python/tvm/contrib/graph_runtime.py

```diff
@@ -39,9 +38,113 @@ def create(graph_json_str, libmod, ctx):
         hmod = rpc_base._ModuleHandle(libmod)
         fcreate = ctx._rpc_sess.get_function("tvm.graph_runtime.remote_create")
         device_type = device_type % rpc_base.RPC_SESS_MASK
-        return GraphModule(fcreate(graph_json_str, hmod, device_type, device_id), ctx)
+        return GraphModule(fcreate(graph_json_str, hmod, device_type,
+                                   device_id), ctx)
+
     fcreate = get_global_func("tvm.graph_runtime.create")
-    return GraphModule(fcreate(graph_json_str, libmod, device_type, device_id), ctx)
+    return GraphModule(fcreate(graph_json_str, libmod, device_type,
+                               device_id), ctx)
+
+
+def _create_heterogeneous(graph_json_str, libmod_ctx, local_ctx):
```

Comment:
Do we need two versions of create functions here ?

By tagging device context to each node and storage plan I see only the below changes
- Storage allocations on multiple devices based on the context.
-  A special operator "__copy" for a direct copy in run function.

Advice to keep implementation to be flexible for more than two devices (it's possible to have multiple devices  Qualcomm Snapdragon:  GPU+DSP+CPU and HiSilicon: GPU+NPU+CPU). Also supporting multiple device ```mblob``` is feasible with current architecture.

I feel users should be unaware of this heterogeneous execution. At most we could have a on/off switch at the compiler front end level to allow CPUFallback to be built into the module or not.

Please let me know if I miss anything here.

Reply:
@srkreddy1238 @tmoreau89 Thanks for your comments. The current implementations works for more than 2 devices. These two `create` functions are transparent to users. They only need to call `graph_runtime.create()`, but they also need to pass in multiple binaries currently. I will look into Tianqi's suggestion above these days when I have time so that we can only have one library file.